### PR TITLE
fix(registry): preserve mirror URL path prefix when resolving/pulling backend images

### PR DESCRIPTION
**Pull request by ilopezluna**
## Context

Follow-up to #974, which routed the llama.cpp backend tag→digest resolution through the mirror-aware containerd resolver. While diagnosing [CSESC-1468](https://docker.atlassian.net/browse/CSESC-1468) on the latest logs, the backend pull still failed with a `403 Forbidden` HEAD to `registry-1.docker.io`, even though the mirror was configured and active.

Root cause: `registryutil.RegistryHosts` **dropped the path component** of the configured mirror URL and hardcoded the Registry v2 API root to `/v2` on the bare host. A mirror served under a path prefix — notably a JFrog Artifactory repository path like `/artifactory/api/docker/<repo>` — was therefore queried at the host root (`<host>/v2/...`), missed, and the resolver fell through to the upstream `registry-1.docker.io`, which 403s behind the corporate proxy. The upstream 403 masked the real (silent) mirror miss.

This is independent of #974 and affects **both** code paths that build hosts via `RegistryHosts`: tag resolution (`ResolveDigest`) and the image pull (`PullPlatform`).

## Changes

- **Preserve the mirror path** — `RegistryHosts` now builds `Path = strings.TrimRight(u.Path, "/") + "/v2"` instead of forcing `/v2` on the bare host. A mirror `https://host/artifactory/api/docker/<repo>` now resolves manifests at `https://host/artifactory/api/docker/<repo>/v2/<name>/manifests/<ref>`.
- **Robust scheme-less parsing** — a mirror given without a scheme (e.g. `host:5000/path`) parses with an empty `Host` and the whole value in `Path`; re-parse it as `https://…` so host and path are separated correctly (the previous code put the entire string into `Host`).
- Mirrors with no path keep the prior `/v2` behavior.

## Tests

- `pkg/internal/registryutil/mirrors_test.go` (new) — table-driven coverage of the resulting `Host`/`Scheme`/`Path` across mirror forms (path prefix, trailing slash, no path, scheme-less with/without path, `http` scheme), plus the upstream fallback ordering and the non-Docker-Hub bypass.
- `pkg/internal/dockerhub/download_test.go` — `TestResolveDigest_UsesMirrorPathPrefix`: an `httptest` registry that serves **only** under `/artifactory/docker/v2` (404 at the host root), so it fails if the path-preserving behavior regresses.

## Verification

- `go build ./...`, `go vet ./pkg/internal/registryutil/... ./pkg/internal/dockerhub/...`
- `go test ./pkg/internal/registryutil/... ./pkg/internal/dockerhub/...` (goleak clean)

## Note

This fix is necessary but not sufficient on its own for the CSESC-1468 customer: they still need to identify the exact URL at which their Artifactory exposes a compatible Docker Registry v2 API (their `/artifactory/docker/v2/` returned 404). Once that base URL is configured as the mirror, this change ensures Model Runner honors its path.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[CSESC-1468]: https://docker.atlassian.net/browse/CSESC-1468?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/internal/registryutil/mirrors.go [hunk 1]
```diff
@@ -39,18 +40,27 @@ func RegistryHosts(mirrors []string, authorizer docker.Authorizer, client *http.
...
+			// A mirror given without a scheme (e.g. "host:5000/path") parses with
+			// an empty Host and the whole value in Path. Re-parse it as https so
+			// the host and any path prefix are separated correctly.
+			if u.Host == "" {
+				u, err = url.Parse("https://" + mirror)
+				if err != nil {
+					slog.Warn("skipping invalid registry mirror", "mirror", mirror, "error", err)
+					continue
+				}
 			}
```
[the remark this answers is not in the document]
Comment (ilopezluna) [reply pkg/internal/registryutil/mirrors.go]: Good catch — fixed in 95013e29f. Now the scheme is prepended before the first `url.Parse` when the mirror has no `://`, so `127.0.0.1:5000` no longer errors on the colon, and we skip only when the resulting host is empty.

File: pkg/internal/registryutil/mirrors_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+			wantScheme: "https",
+			wantPath:   "/artifactory/docker/v2",
+		},
+		{
+			name:       "no scheme bare host",
+			mirror:     "mirror.example.com:5000",
+			wantHost:   "mirror.example.com:5000",
+			wantScheme: "https",
+			wantPath:   "/v2",
+		},
```
[the remark this answers is not in the document]
Comment (ilopezluna) [reply pkg/internal/registryutil/mirrors_test.go:82]: Added `127.0.0.1:5000` and `127.0.0.1:5000/artifactory/docker` test cases in 95013e29f.

File: pkg/internal/registryutil/mirrors.go [hunk 3]
```diff
@@ -34,23 +35,34 @@ func RegistryHosts(mirrors []string, authorizer docker.Authorizer, client *http.
...
-				scheme = "https"
+			if u.Host == "" {
+				slog.Warn("skipping invalid registry mirror", "mirror", mirror, "error", "empty host")
+				continue
 			}
+			// Preserve any path prefix on the mirror (e.g. a JFrog Artifactory
+			// repository path "/artifactory/api/docker/<repo>") and append the
+			// Registry v2 API root. Without this, a mirror configured with a path
+			// would be queried at the host root and fail.
+			path := strings.TrimRight(u.Path, "/") + "/v2"
```
[the remark this answers is not in the document]
Comment (ilopezluna) [reply pkg/internal/registryutil/mirrors.go]: Good catch — guarded in 5d1fad7f9. We now only append `/v2` when the configured path doesn't already end with it (via `strings.HasSuffix`), and added test cases for `.../repo/v2` and `/v2/` so the suffix isn't doubled.

**Comment by ilopezluna** [comment]
@sourcery-ai review

**Comment by ilopezluna** [comment]
Thanks @sourcery-ai — extracted the mirror URL normalization (scheme defaulting, path-prefix preservation, `/v2` suffix handling) into a documented `parseMirror` helper in db8f0c47d, keeping the `RegistryHosts` loop small. No behavior change; existing tests still cover the edge cases.

**Comment by ilopezluna** [comment]
@sourcery-ai review

**Review by doringeman — APPROVED** [review]
